lib: crc: add CRC-32K/4.2

JPHutchins opened the pull request:
This adds the best HD=4 CRC32 polynomial. The discovery is the result of research by Philip Koopman of Carnegie Mellon University, and is well documented at
https://users.ece.cmu.edu/~koopman/crc/.

The user is given the option of trading 1024B of RAM to double the execution speed. The unit tests are parameterized with this KConfig option.

Discussion: https://github.com/zephyrproject-rtos/zephyr/discussions/80464

JPHutchins commented:
stale

pdgendt commented on include/zephyr/sys/crc.h:
```diff
@@ -263,6 +265,32 @@ uint32_t crc32_ieee_update(uint32_t crc, const uint8_t *data, size_t len);
+ *
+ * @return CRC32 value.
+ */
+uint32_t crc32_k_4_2_update(uint32_t crc, uint8_t const *data, size_t len);
```
Put `const` modifiers before the type, to keep it consistent. (Here and other new code)

pdgendt commented:
@JPHutchins any update on this?

JPHutchins commented:
> @JPHutchins any update on this?

Thank you for the reminder, I lost track of this!

JPHutchins committed:
lib: crc: add CRC-32K/4.2

This adds the best HD=4 CRC32 polynomial. The discovery
is the result of research by Philip Koopman of Carnegie
Mellon University, and is well documented at
https://users.ece.cmu.edu/~koopman/crc/.

The user is given the option of trading 1024B of RAM to
improve the execution speed. The unit tests are parameterized
with this KConfig option.

Signed-off-by: JP Hutchins <jp@intercreate.io>

kartben reviewed (commented):
Can you please follow-up with an update to the release notes for Zephyr 4.2 documenting the new APIs? Thanks!
